### PR TITLE
Add first draft of `first_factor` paper

### PR DESCRIPTION
**Pull request by chiphogg**
For now, I'm assuming we'll get the paper number P2357.

The major known missing piece is references and citations.  However, I expect we can land this PR with whatever minimal narrative changes we need, and iterate from there.

### CODE REVIEW

**Review by mpusz — COMMENTED** [review]
Nice paper! :-)

I would only add a small chapter about the implementability that would state that a standard library can't implement it as a regular `constexpr` functions. If it was possible we would have done already 😉 There should be some hint in the text stating that some intrinsic or "magic" compiler support might be needed to implement it efficiently without the "endless" recursion.

File: src/3133R0_first_factor.md [hunk 1]
```diff
@@ -196,6 +196,24 @@ There may be other overloads besides the above for `std::uint64_t`, as we'll dis
...
+
+Pessimistically, we'll assume that there do exist inputs that will exceed the compiler's preferred
+iteration limits.  If that happens, the compiler will need to circumvent these limits for
+`std::first_factor` in order to be compliant with the standard.
+
+We are open to suggestion as to the mechanism.  Perhaps the compiler can provide directives to
+disable these limits for _individual blocks of code_.  These directives could be either private (for
+compiler-internal use cases only), or public (for end users as well).  In any case, the most
+important point for this present proposal is that `std::first_factor` must be able to produce
+a result for every input at compile time.
```
Comment (mpusz) [inline src/3133R0_first_factor.md:214]: We do not have to suggest the mechanism. Each compiler can choose what they like. Also, suggesting that we might need to change the iteration limits in parts of the code might not be most lucky also.

I would rather write that the library function, at least when evaluated at compile time, may need some help from the compiler internals (e.g., this is [common for some type-traits](https://stackoverflow.com/questions/49901309/is-it-possible-to-have-a-library-implementation-of-is-trivially-constructible-wi)).

File: src/3133R0_first_factor.md [hunk 2]
```diff
@@ -344,10 +362,20 @@ library meaningfully harder to implement.  It would also mean that other project
...
+
 # Conclusion
 
 Integer factorization is perhaps the most basic and useful operation in all of number theory.  While
 basic implementations are easy to write, robust and efficient implementations are tricky and subtle.
-Providing a high quality implementation in the C++ standard library would unblock at least a future
-standard units and quantities library.  Given the wide diversity of use cases for integer
+If results are required at compile time, the difficulty increases even more, because compilers limit
+the number of algorithmic iterations.  In fact, there's a real risk that a `constexpr` function that
+produces results for every 64-bit input is even possible for end user code!
```
Comment (mpusz) [inline src/3133R0_first_factor.md:375]: Did you mean?

```suggestion
produces results for every 64-bit input is even impossible for end user code!
```

File: src/3133R0_first_factor.md [hunk 3]
```diff
@@ -344,10 +362,20 @@ library meaningfully harder to implement.  It would also mean that other project
...
+
+Providing a high quality implementation in the C++ standard library would make a future standard
+units and quantities library much easier to write, and may even be a hard blocker for it.  But
```
Comment (mpusz) [inline src/3133R0_first_factor.md:378]: I do not understand the end of this sentence 😢

**Review by mpusz — CHANGES_REQUESTED** [review]

File: src/3133R0_first_factor.md (hunk 2 above)
Comment (chiphogg) [reply src/3133R0_first_factor.md:375]: I meant "not even possible".  Nice catch!

File: src/3133R0_first_factor.md (hunk 3 above)
Comment (chiphogg) [reply src/3133R0_first_factor.md]: Yeah, that was _really_ awkward.  I've got a rephrasing up now.  I'm sure it's better than it was --- but is it good _enough_?

**Review by mpusz — APPROVED** [review]
Nice! :-)

**Pull request merged**